src/Dictionaries/HashedDictionaryCollectionType.h

```diff
@@ -0,0 +1,218 @@
+#pragma once
+
+#include <Dictionaries/IDictionary.h>
+#include <Common/HashTable/PackedHashMap.h>
+#include <Common/HashTable/HashMap.h>
+#include <Common/HashTable/HashSet.h>
+#include <sparsehash/sparse_hash_map>
+#include <sparsehash/sparse_hash_set>
+#include <type_traits>
+
+namespace DB
+{
+
+/// HashMap with packed structure is better then google::sparse_hash_map if the
+/// <K, V> pair is small, for the sizeof(std::pair<K, V>) == 16, RSS for hash
+/// table with 1e9 elements will be:
+///
+/// - google::sparse_hash_map             : 26GiB
+/// - HashMap                             : 35GiB
+/// - PackedHashMap                       : 22GiB
+/// - google::sparse_hash_map<packed_pair>: 17GiB
+///
+/// Also note here sizeof(std::pair<>) was used since google::sparse_hash_map
+/// uses it to store <K, V>, yes we can modify google::sparse_hash_map to work
+/// with packed analog of std::pair, but the allocator overhead is still
+/// significant, because of tons of reallocations (and those cannot be solved
+/// with reserve() due to some internals of google::sparse_hash_map) and poor
+/// jemalloc support of such pattern, which results in 33% fragmentation (in
+/// comparison with glibc).
+///
+/// Plus since google::sparse_hash_map cannot use packed structure, it will
+/// have the same memory footprint for everything from UInt8 to UInt64 values
+/// and so on.
+///
+/// Returns true hen google::sparse_hash_map should be used, otherwise
+/// PackedHashMap should be used instead.
+template <typename K, typename V>
+constexpr bool useSparseHashForHashedDictionary()
+{
+    return sizeof(PackedPairNoInit<K, V>) > 16;
+}
+
+/// Grower with custom fill limit/load factor (instead of default 50%).
+///
+/// It turns out that HashMap can outperform google::sparse_hash_map in case of
+/// the structure size of not big, in terms of speed *and* memory. Even 99% of
+/// max load factor was faster then google::sparse_hash_map in my simple tests
+/// (1e9 UInt64 keys with UInt16 values, randomly distributed).
+///
+/// And not to mention very high allocator memory fragmentation in
+/// google::sparse_hash_map.
+///
+/// Based on HashTableGrowerWithPrecalculation
+class alignas(64) HashTableGrowerWithMaxLoadFactor
```

Comment:
> (cosmetic: could use a name that positions this class better within the existing system of growers, e.g. HashTableGrowerWithPrecalculationAndMaxLoadFactor

I thought that even for ClickHouse it will be too long, but if you have this mind I will update the code.

>Please also add a comment to HashTableGrower that it assumes a max load factor of 0.5.).

Ok

>I actually also think that HashTable.h would be a more fitting place for HashTableGrowerWithMaxLoadFactor (right after the existing growers).

Let's keep it here for now, since changes in `HashTable.h` requires tons of modules to recomple, once it will be stable it could be moved there.